src/edge/tests/test_io.py

```diff
@@ -70,8 +90,125 @@ def test_outputs_gff_handles_negative_strand(self):
         # be aware of the tabs in the string below
         expected = """##gff-version 3
 ##sequence-region Bar 1 19
-Bar\tfeature\tgene\t2\t2\t.\t-\t.\tname=A1
-Bar\tfeature\tgene\t9\t15\t.\t-\t.\tname=A1
+Bar\tfeature\tgene\t2\t2\t.\t+\t.\tname=A1%%5B1:1%%5D
+Bar\tfeature\tgene\t9\t15\t.\t+\t.\tname=A1%%5B2:8%%5D
+##FASTA
+>Bar <unknown description>
+%s
+""" % (fragment.sequence,)
+        self.assertEquals(expected, gff)
+
+    def test_outputs_gff_with_hierarchical_annotations(self):
+        fragment = self.fragment.indexed_fragment()
+        fragment.annotate(2, 9, 'A1', 'gene', 1)
+        fragment.annotate(2, 9, 'A1', 'mRNA', 1)
+        fragment.annotate(2, 9, 'A1', 'CDS', 1)
+
+        with tempfile.NamedTemporaryFile(mode='w+', delete=False) as f:
+            f.close()
+            IO(self.genome).to_gff(f.name)
+            h = open(f.name, 'r')
+            gff = h.read()
+            h.close()
+            os.unlink(f.name)
+
+        # be aware of the tabs in the string below
+        expected = """##gff-version 3
+##sequence-region Bar 1 13
+Bar\tfeature\tgene\t2\t9\t.\t+\t.\tname=A1
+Bar\tfeature\tmRNA\t2\t9\t.\t+\t.\tname=A1
+Bar\tfeature\tCDS\t2\t9\t.\t+\t0\tname=A1
+##FASTA
+>Bar <unknown description>
+%s
+""" % (fragment.sequence,)
+        self.assertEquals(expected, gff)
+
+    def test_outputs_gff_features_with_null_phase(self):
+        fragment = self.fragment.indexed_fragment()
+        fragment.annotate(2, 9, 'A1', 'CDS', 1, qualifiers=dict(phase=None))
+
+        with tempfile.NamedTemporaryFile(mode='w+', delete=False) as f:
+            f.close()
+            IO(self.genome).to_gff(f.name)
+            h = open(f.name, 'r')
+            gff = h.read()
+            h.close()
+            os.unlink(f.name)
+
+        # be aware of the tabs in the string below
+        expected = """##gff-version 3
```

Comment:
is there anyway to make these gff fixtures look a bit nicer? the indentation makes the python file quite ugly.

Reply:
I think there is no easy way to make this prettier, I believe the way I did in the duplo pcr file generation was to put this as a separate fixture file but given the amount of different small gff files here it makes it harder to manage. I'm good with leaving it this way for now.